contracts/asset-proxy/contracts/src/ERC1155Proxy.sol

```diff
@@ -0,0 +1,268 @@
+            // assetData Encoded byte array.
+            // from Address to transfer asset from.
+            // to Address to transfer asset to.
+            // amount Amount of asset to transfer.
+            // bytes4(keccak256("transferFrom(bytes,address,address,uint256)")) = 0xa85e59e4
+            if eq(selector, 0xa85e59e400000000000000000000000000000000000000000000000000000000) {
+                
+                // To lookup a value in a mapping, we load from the storage location keccak256(k, p),
+                // where k is the key left padded to 32 bytes and p is the storage slot
+                mstore(0, caller)
+                mstore(32, authorized_slot)
+
+                 // Revert if authorized[msg.sender] == false
+                if iszero(sload(keccak256(0, 64))) {
+                    // Revert with `Error("SENDER_NOT_AUTHORIZED")`
+                    mstore(0, 0x08c379a000000000000000000000000000000000000000000000000000000000)
+                    mstore(32, 0x0000002000000000000000000000000000000000000000000000000000000000)
+                    mstore(64, 0x0000001553454e4445525f4e4f545f415554484f52495a454400000000000000)
+                    mstore(96, 0)
+                    revert(0, 100)
+                }
+
+                // Construct Table #3 in memory, starting at memory offset 0.
+                // The algorithm below maps asset data from Table #1 and Table #2 to Table #3, while
+                // scaling the `tokenValues` (Table #2) by `amount` (Table #1). Once Table #3 has
+                // been constructed in memory, the destination erc1155 contract is called using this 
+                // as its calldata. This process is divided into four steps, below.
+
+                ////////// STEP 1/4 //////////
+                // Map relevant fields from assetData (Table #2) into memory (Table #3)
+                // The Contents column of Table #2 is the same as Table #3,
+                // beginning from parameter 3 - `offset to tokenIds (*)`
+                // The offsets in these rows are offset by 32 bytes in Table #3.
+                // Strategy:
+                // 1. Copy the assetData into memory at offset 32
+                // 2. Increment by 32 the offsets to `tokenIds`, `tokenValues`, and `callbackData`
+
+                // Load offset to `assetData`
+                let assetDataOffset := calldataload(4)
+
+                // Load length in bytes of `assetData`, computed by:
+                // 4 (function selector)
+                // + assetDataOffset
+                let assetDataLength := calldataload(add(4, assetDataOffset))
+
+                // This corresponds to the beginning of the Data Area for Table #3.
+                // Computed by:
+                // 4 (function selector)
+                // + assetDataOffset
+                // + 32 (length of assetData)
+                calldatacopy(32, add(36, assetDataOffset), assetDataLength)
```

Comment:
You're probably right, but why not `calldatacopy(68, add(36, assetDataOffset), sub(assetDataLength, 36))`?

Reply:
Good question - this line copies the entire assetData into memory at offset 32. Memory allocation looks like this:
```
32    :    assetProxyId 
36    :    address of ERC1155 contract
68    :    offset to tokenIds
...
```

This is later transformed into:
```
00    :    batchTransferFrom selector
04    :    from address
36    :    to address
68    :    offset to tokenIds
...
```

If we replaced this line with `calldatacopy(68, add(36, assetDataOffset), sub(assetDataLength, 36))` then we would have the following allocation:
```
68    :    assetProxyId 
72    :    address of ERC1155 contract
104   :    offset to tokenIds
...
```
And it would truncate 36 bytes at the tail end of the asset data.

Note that `assetData` is a byte array so it is abi encoded as:
```
0   : length
32  : contents
```

Reply:
Thanks! Super appreciate the walkthrough.